### DISCUSSION TITLE
Improve HLIL for signed non-negative compare

### DISCUSSION THREAD

**Discussion by gixslayer** [Vector35/binaryninja-api#2892] [Ideas] [answered]
Tested on version `3.0.3233`. I have a simple piece of x86_64 code that essentially performs the following:

```.c
int32_t value;
// code that assigns to and uses value
return value >= 0;
```

This greater or equal to zero check is implemented as:

```.asm
; load value into eax
not eax
shr eax, 0x1f
; return
```

So effectively it just isolates the sign bit and checks that.
This is correctly shown in the disassembly and BNIL, but the HLIL also shows this in a very verbose manner:

```.c
rax_25 = (not.d(value) u>> 0x1f).b
```

It would be nice if the HLIL would display it as something similar to `value s>= 0`. Not sure if this is just an edge case, but it feels common enough to warrant some QoL readability improvements for HLIL.

**Comment by plafosse** [comment, marked as answer]
Thanks for the question, we moved this to an issue.
